oauth2.md

```diff
@@ -0,0 +1,165 @@
+- `redirect_uri`: One of the registered `redirect_uri` from step 1
+- `state`: a random string generated by your app to prevent CSRF attacks
+
+```html
+<!--
+Example Values:
+- response_type: code
+- client_id: b6084afa-e9ee-4eb4-8220-338be0c12374
+- redirect_uri: https://your-app.example.com/callback
+- state: v6glrJn3gf3qL4rPFLBB
+-->
+
+<a href="https://oauth.mocaverse.xyz/api/oauth/auth?response_type=code&client_id=b6084afa-e9ee-4eb4-8220-338be0c12374&redirect_uri=https%3A%2F%2Fyour-app.example.com%2Fcallback&state=v6glrJn3gf3qL4rPFLBB">
+  Login with Mocaverse
+</a>
+```
+
+If the `client_id` and `redirect_uri` parameters are correct, any errors that occur will be sent via HTTP 302 redirect to the specified `redirect_uri`.
+
+```
+# response_type: some_invalid_type
+# redirect_uri: https://your-app.example.com/callback
+# state: dK6fRBNVY1HM2GQ4aBwM
+
+< HTTP/1.1 302 Found
+Found. Redirecting to https://your-app.example.com/callback?error=unsupported_response_type&error_description=only+supports%3A+code&state=dK6fRBNVY1HM2GQ4aBwM
+```
+
+If any of the `client_id` and `redirect_uri` are invalid, the error will be provided as a JSON payload as HTTP 400.
+
+```
+< HTTP/1.1 400 Bad Request
+{
+  "error": "invalid_request",
+  "error_description": "invalid client_id",
+  "state": "dK6fRBNVY1HM2GQ4aBwM"
+}
+```
+
+### 3. User Authorization
+
+The user will be redirected to log in and authorize your app on Mocaverse's site.
+
+### 4. Authorization Code Redirect Callback
+
+If the user approves the authorization request, they will be redirected back to your `redirect_uri` containing a one-time use authorization `code` and the provided `state` from step 2.
+
+```
+< HTTP/1.1 302 Found
+Found. Redirecting to https://your-app.example.com/callback?code=602cbbe1-60dd-482f-b0d8-329f5f1254c7&state=v6glrJn3gf3qL4rPFLBB
+```
+
+If the user denies the authorization request, the user will be redirected to your `redirect_uri` containing an `access_denied` error and the provided `state` from step 2.
+
+```
+< HTTP/1.1 302 Found
+Found. Redirecting to https://your-app.example.com/callback?error=access_denied&state=E0BXkRkKnvqMiDdYC8MW
+```
+
+Important: Validating the `state` if the value matches with the initial request during the User Authorization Request is strongly recommended.
+
+### 5. Access Token Request
+
+#### Client Secret Authentication
+
+Request the `access_token` via `HTTP POST https://oauth.mocaverse.xyz/api/oauth/token` with the following `application/x-www-form-urlencoded` parameters:
+
+- `grant_type`: must be `authorization_code`
+- `client_id`: must be `urn:ietf:params:oauth:client-assertion-type:jwt-bearer`
+- `client_secret`: send the private key related public key provided to Moca from Application Registration. PEM headers can be discarded.
+- `code`: the authorization code received from step 4
+- `redirect_uri` must match the same value used during **Authorization Request**
+
+#### JWT Client Authentication
```

Comment:
Not needed right

Reply:
Yes. But kept it as the code can still accommodate jwt client authentication. I'll update the headers, to make it more clear.